homeassistant/components/numato/binary_sensor.py

```diff
@@ -0,0 +1,121 @@
+"""Binary sensor platform integration for Numato USB GPIO expanders."""
+from functools import partial
+import logging
+
+from numato_gpio import NumatoGpioError
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.const import DEVICE_DEFAULT_NAME
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect, dispatcher_send
+
+from . import (
+    CONF_BINARY_SENSORS,
+    CONF_DEVICES,
+    CONF_ID,
+    CONF_INVERT_LOGIC,
+    CONF_PORTS,
+    DATA_API,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+NUMATO_SIGNAL = "numato_signal_{}_{}"
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the configured Numato USB GPIO binary sensor ports."""
+    if discovery_info is None:
+        return
+
+    api = hass.data[DOMAIN][DATA_API]
+    binary_sensors = []
+    devices = hass.data[DOMAIN][CONF_DEVICES]
+    for device in [d for d in devices if CONF_BINARY_SENSORS in d]:
+        device_id = device[CONF_ID]
+        platform = device[CONF_BINARY_SENSORS]
+        invert_logic = platform[CONF_INVERT_LOGIC]
+        ports = platform[CONF_PORTS]
+        for port, port_name in ports.items():
+            try:
+
+                def read_gpio(device_id, port, level, name=port_name):
```

Comment:
Move this up outside the loops. We use partial to set the arguments we get from the loops.

Reply:
Real tired... I don't even need the port name here...
Thanks for the thorough review!